primitives/src/targeting/eval.rs

```diff
@@ -411,6 +722,102 @@ fn eval(input: &Input, output: &mut Output, rule: &Rule) -> Result<Option<Value>
 
             Some(Value::Bool(a.iter().any(|x| b.contains(x))))
         }
+        Function::In(first_rule, second_rule) => {
+            let a = eval(input, output, first_rule)?.ok_or(Error::TypeError)?;
+            let b = eval(input, output, second_rule)?
+                .ok_or(Error::TypeError)?
+                .try_array()?;
+
+            Some(Value::Bool(b.contains(&a)))
+        }
+        Function::NotIn(first_rule, second_rule) => {
+            let a = eval(input, output, first_rule)?.ok_or(Error::TypeError)?;
+            let b = eval(input, output, second_rule)?
+                .ok_or(Error::TypeError)?
+                .try_array()?;
+
+            Some(Value::Bool(!b.contains(&a)))
+        }
+        Function::Between(first_rule, second_rule, third_rule) => {
+            let first_eval = first_rule.eval(input, output)?.ok_or(Error::TypeError)?;
+            let second_eval = second_rule.eval(input, output)?.ok_or(Error::TypeError)?;
+            let third_eval = third_rule.eval(input, output)?.ok_or(Error::TypeError)?;
+
+            let value = BigNum::try_from(first_eval)?;
+            let start = BigNum::try_from(second_eval)?;
+            let end = BigNum::try_from(third_eval)?;
+
+            Some(Value::Bool(value.ge(&start) && value.le(&end)))
+        }
+        Function::At(first_rule, second_rule) => {
+            let mut first_eval = first_rule
+                .eval(input, output)?
+                .ok_or(Error::TypeError)?
+                .try_array()?;
+            let second_eval = second_rule
+                .eval(input, output)?
+                .ok_or(Error::TypeError)?
+                .try_number()?
+                .as_u64()
+                .ok_or(Error::TypeError)?;
+            let index = usize::try_from(second_eval).unwrap();
+            let value = if first_eval.get(index).is_none() {
+                None
+            } else {
+                Some(first_eval.swap_remove(index))
+            };
+            let value = value.ok_or(Error::TypeError)?;
```

Comment:
Maybe add a more descriptive error here?

Reply:
Not really needed at this point.
However you can simply leave the if-else at the end and drop the:

```rust
let value = value.ok_or(Error::TypeError)?;
Some(value)
```
with smth like:
```rust
if first_eval.get(index).is_none() {
                return Err(Error::TypeError);
            } else {
                Some(first_eval.swap_remove(index))
            }
```